fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -64,24 +66,19 @@ where
                 FvmQueryRet::Call(ret)
             }
             FvmQuery::EstimateGas(mut msg) => {
-                // TODO: Figure out how to relate the token balance to a gas limit.
-                //       Do we look at `state.state_params.base_fee`? What about `gas_premium` and `gas_limit`?
-
-                // XXX: This value is for Filecoin, and it's not even used by the FVM, but at least it should not have a problem with it.
+                // Setting BlockGasLimit as initial limit for gas estimation
                 msg.gas_limit = BLOCK_GAS_LIMIT;
 
-                // TODO: This actually fails if the caller doesn't have enough gas.
-                //       Should we modify the state tree up front to give it more?
-                let apply_ret = state.call(*msg)?;
+                // Populate gas message parameters.
+                // If message fails with BLOCK_GAS_LIMIT as the message gas limit
+                // it means that there is an error in the execution of the message
+                // we could optionally propagate that error if needed as done in Lotus.
+                // (but why would you want to estimate the gas of a message that can't
+                // be executed?)
+                let mut msg = self.estimate_gassed_msg(&state, msg)?;
```

Comment:
I don't understand: this method will put in the exact gas used with a 25% over estimation in the message. Why should we have to perform a `gas_search` after this, which further increases it by 20% in each iteration? Why would it suddenly fail to execute on the same exact state, when gas is a deterministic value?

Reply:
So the idea here is that if the overestimation fails (i.e. we haven´t set enough gas yet), we may need to search for a bigger value.

I reduced the overestimation and steps to 5% now, and now that I removed the `gas_fee_cap` it may always get it at the first attempt, but in case this changes in the future I think we should keep the `gas_search`.